### ISSUE TITLE
Expand/Collapse All Content Toggle within WET's tabbed interface

### ISSUE THREAD

**Issue by olegk1** [wet-boew/wet-boew#6416]
Hi there! 
Is it possible to use Expand/Collapse All Content Toggle within WET's tabbed interface and also within nested collapsible bars on the same page. I re-used WET's toggle example code applying two CSS classes: for parent container and for selector to open/close collapsible bars, but it didn't work. To be precise, it didn't work at all within tabbed interface. As for nested collapsible bars, it worked fine on Windows, but produced inconsistent results on Mac's browsers. I'm wondering if WET toggle is designed to be used within tabbed interface and within nested collapsible bars? Thank you!

**Comment by olegk1** [comment]
Hi Laurent: 
I've just emailed you the simplified version of the code I'm using as I cannot attach it here for some reason. Hopefully it makes sense. Thanks!

**Comment by pjackson28** [comment]
@olegk1 You can include code using three backticks (a.k.a. grave accent) followed by an optional language (e.g., html, css, js, scss) to open the code block and three backticks to close the code block. An example for JavaScript is shown in the "Syntax highlighting" section of https://guides.github.com/features/mastering-markdown/ (they use the expanded version of js but the abbreviated version works fine as well).

**Comment by olegk1** [comment]
Thanks a lot, Paul! 

Here's two snippets of my code where I use WET toggle.

```
<!—Toggle inside nested collapsible bars within tabbed interface code //
===================================
<!—main collapsible bar START //
<details>
<summary>
<h2>Main collapsible bar TITLE</h2>
</summary>
<div class="wb-tabs ">
	<div class="tabpanels">
<!—first tab //
    		<details open>
		    <summary>
                                   <h3>2013-14 tab</h3>
                               </summary>
<div class="parent-boxTabs">
<p><a href="#" class="wb-toggle" data-toggle="{&quot;selector&quot;: &quot;. child-boxTabs &quot;,&quot;parent&quot;: &quot;. parent-boxTabs &quot;, &quot;print&quot;: &quot;on&quot;}">OPEN ALL</a>&nbsp;&nbsp;|&nbsp;&nbsp;<a href="#" class="wb-toggle" data-toggle="{&quot;selector&quot;: &quot;. child-boxTabs &quot;,&quot;parent&quot;: &quot;.parent-boxTab &quot;,&quot;print&quot;: &quot;off&quot;}">CLOSE ALL</a></p>
<div class=”child-boxTabs”>
<details>
<summary>
<h4>Nested collapsible bar TITLE 1</h4>
</summary>
<div>
Nested collapsible bar TITLE 1 content
</div>
</details>
<details>
<summary>
<h4>Nested collapsible bar TITLE 2</h4>
</summary>
<div>
Nested collapsible bar TITLE2 content
</div>
</details>
</div>
</details>
<!—second tab //
<details >
		    <summary>
                                   <h3>2012-13 tab</h3>
                               </summary>
<div class="parent-boxTabs">
<p><a href="#" class="wb-toggle" data-toggle="{&quot;selector&quot;: &quot;. child-boxTabs &quot;,&quot;parent&quot;: &quot;. parent-boxTabs &quot;, &quot;print&quot;: &quot;on&quot;}">OPEN ALL</a>&nbsp;&nbsp;|&nbsp;&nbsp;<a href="#" class="wb-toggle" data-toggle="{&quot;selector&quot;: &quot;. child-boxTabs &quot;,&quot;parent&quot;: &quot;.parent-boxTab &quot;,&quot;print&quot;: &quot;off&quot;}">CLOSE ALL</a></p>
<div class=”child-boxTabs”>
<details>
<summary>
<h4>Nested collapsible bar TITLE 1</h4>
</summary>
<div>
Nested collapsible bar TITLE 1 content
</div>
</details>
<details>
<summary>
<h4>Nested collapsible bar TITLE 2</h4>
</summary>
<div>
Nested collapsible bar TITLE2 content
</div>
</details>
</div>
</details>
</div>
</div>
</details>
<!—main collapsible bar END //

```

```
<!—Toggle inside nested collapsible bars within main collapsible bar code //
===================================
<!—main collapsible bar START //
<details>
<summary>
<h2> Main collapsible bar TITLE </h2>
</summary>
<div class="parent-box">
<p><a href="#" class="wb-toggle" data-toggle="{&quot;selector&quot;: &quot;. child-box &quot;,&quot;parent&quot;: &quot;. parent-box &quot;, &quot;print&quot;: &quot;on&quot;}">OPEN ALL</a>&nbsp;&nbsp;|&nbsp;&nbsp;<a href="#" class="wb-toggle" data-toggle="{&quot;selector&quot;: &quot;. child-box &quot;,&quot;parent&quot;: &quot;. parent-box &quot;,&quot;print&quot;: &quot;off&quot;}">CLOSE ALL</a></p>
<details class="child-box">
<summary>
<h3>Nested Collapsible Bar TITLE 1</h3>
</summary>
<div class="wb-tabs ">
   <div class="tabpanels">
        <details open="open">
           <summary>
               <h3> 2014 tab</h3>
            </summary>
        <div>
              Collapsible Bar TITLE 1: 2014 tab content
       </div>
     </details>
   <details>
           <summary>
               <h3> 2013 tab</h3>
            </summary>
        <div>
              Collapsible Bar TITLE 1: 2013 tab content
       </div>
     </details>
</div>
</div>
</details>
<!--// second collapsible bar
<details class="child-box">
<summary>
<h3>Nested Collapsible Bar TITLE 2</h3>
</summary>
<div class="wb-tabs ">
   <div class="tabpanels">
        <details open="open">
           <summary>
               <h3> 2014 tab</h3>
            </summary>
        <div>
              Collapsible Bar TITLE 2: 2014 tab content
       </div>
     </details>
   <details>
           <summary>
               <h3> 2013 tab</h3>
            </summary>
        <div>
              Collapsible Bar TITLE 2: 2013 tab content
       </div>
     </details>
</div>
</div>
</details>
</div>
</details>
<!—main collapsible bar END //


```

**Comment by pjackson28** [comment]
@olegk1 You have a lot of markup issues including the following:
* Missing closing div elements
* Using smart quotes (”) instead of normal quotes (") for some attributes
* data-toggle attribute is coded incorrectly including:
 * Attribute value needs to be and single quotes and included values in double quotes (as per what is required for JSON)
 * Using "print" instead of "type"
 * "selector" needs to point to the details elements, not the parent container
 * Reusing parent for different tabs (each parent needs to be different)
 * Using an "a" element when you should be using a "button" element. If you really want to use an a element then you need to change the href to "javascript:;" (although still should use button).

The following is the cleaned up markup for your first snippet which works fine for me:
```html
<details>
	<summary class="h2">
		<h2>Main collapsible bar TITLE</h2>
	</summary>
	<div class="wb-tabs">
		<div class="tabpanels">
			<details open="open">
				<summary>
					<h3>2013-14 tab</h3>
				</summary>
				<div class="parent-boxTabs">
					<div class="btn-group">
						<button type="button" class="btn btn-default wb-toggle" data-toggle='{"selector": "details", "parent": ".parent-boxTabs", "type": "on"}'>OPEN ALL</button>
						<button type="button" class="btn btn-default wb-toggle" data-toggle='{"selector": "details", "parent": ".parent-boxTabs", "type": "off"}'>CLOSE ALL</button>
					</div>
					<details>
						<summary>
							<h4>Nested collapsible bar TITLE 1 (2013-14 tab)</h4>
						</summary>
						<div>
							Nested collapsible bar TITLE 1 content (2013-14 tab)
						</div>
					</details>
					<details>
						<summary>
							<h4>Nested collapsible bar TITLE 2 (2013-14 tab)</h4>
						</summary>
						<div>
							Nested collapsible bar TITLE2 content (2013-14 tab)
						</div>
					</details>
				</div>
			</details>
			<details >
				<summary>
					<h3>2012-13 tab</h3>
				</summary>
				<div class="parent-boxTabs2">
					<div class="btn-group">
						<button type="button" class="btn btn-default wb-toggle" data-toggle='{"selector": "details", "parent": ".parent-boxTabs2", "type": "on"}'>OPEN ALL</button>
						<button type="button" class="btn btn-default wb-toggle" data-toggle='{"selector": "details", "parent": ".parent-boxTabs2", "type": "off"}'>CLOSE ALL</button>
					</div>
					<details>
						<summary>
							<h4>Nested collapsible bar TITLE 1 (2012-13 tab)</h4>
						</summary>
						<div>
							Nested collapsible bar TITLE 1 content (2012-13 tab)
						</div>
					</details>
					<details>
						<summary>
							<h4>Nested collapsible bar TITLE 2 (2012-13 tab)</h4>
						</summary>
						<div>
							Nested collapsible bar TITLE2 content (2012-13 tab)
						</div>
					</details>
				</div>
			</details>
		</div>
	</div>
</details>
```

**Comment by olegk1** [comment]
Great help, Paul! 
I've modified my code and it works perfect now.

**Comment by olegk1** [comment]
I just wanted to double check something with your code, Paul. Can you tab between tabs in your vesriosn of my code ok? When I tab through my tabbed interface , the focus shift is the following: main collapsible bar1-->(tab)-->nested tab1-->(tab)-->open all button--> (tab)-->close all button-->(tab)-->first nested collapsible bar-->(tab)-->second nested collapsible bar, etc. Instead, how can I implement focus moving on tabbing from "close all button" link to "nested tab2"?  Thanks!

**Comment by pjackson28** [comment]
@olegk1 You can't tab to the other tab until it becomes the active tab. We are following the WAI-ARIA tabpanel design pattern where only the active tabs are in the tabbing order. To go between tabs you are supposed to use the arrow keys (as per the design pattern).

**Comment by olegk1** [comment]
I see. I've got fixed. Thanks again.

**Comment by olegk1** [comment]
Sorry to bug you again, Paul, but could you please try to use keyboard to navigate the cleaned up version you provided? When I use an arrow to move from "2013-14 tab" to "2012-13 tab" and then use Tab to go to "Open All" button, it doesn't work. I need to tab again to get to "Open All" or tabbing doesn't work at all. Thanks!

**Comment by pjackson28** [comment]
There is a bug that inadvertently included the hidden nested summary elements in the tabbing order. As for back to the tabs, I fixed the Ctrl + Up handling which brings focus back to the parent tabs or summary element (when using wb-tabs or wb-toggle accordion). Both fixes are in PR #6426.

**Comment by olegk1** [comment]
Thank you for taking care of this. So i assume the fix is going to be included in the v4.0.10 release? Would you let me know when it is ready for download, please so I can add the functionality back to our pages?

**Comment by pjackson28** [comment]
It will be in v4.0.10 but will soon also be available in the latest development release (WET theme and Canada.ca theme development releases include it now and the other themes should have it in about a half hour).

**Comment by olegk1** [comment]
I see. Thanks!

**Comment by josephbr** [comment]
I updated WET so that it would include tabbing the bug fix and then tried olegk1's page again. However, the tabs completely disappeared. Was the structure/markup for the "Tabbed Interface" changed in that v4.0.10?

**Comment by pjackson28** [comment]
@josephbr You'll have to use the latest development release to get these fixes as v4.0.10 has not been released yet (v4.0.9 won't include the nesting fixes). You can find the development releases after the stable releases on the download page.

**Comment by josephbr** [comment]
I had downloaded it from this page?
https://github.com/wet-boew/wet-boew-dist
Is that not the latest build? I see files there that were updated just hours ago.

**Comment by pjackson28** [comment]
You can find the development releases here: http://wet-boew.github.io/v4.0-ci/docs/versions/dwnld-en.html#v40dev

You could use https://github.com/wet-boew/wet-boew-dist but you would have to pick the correct branch and then click "Download ZIP" (a lot easier just to use the actual download page).

**Comment by josephbr** [comment]
Downloaded the latest development release and the problem remains in that tabs still don't display. If I roll back to an older version (i.e. v4.0.10-development - 2015-01-05), then it's fine but "v4.0.10-development - 2015-01-08" is a no go.

**Comment by pjackson28** [comment]
I just downloaded the latest zip of the WET theme and put the cleaned up markup into one of the pages and it worked fine for me in Firefox, Chrome and IE11. What browser are you using?

**Comment by josephbr** [comment]
I am using Chrome v39.2171.95.

**Comment by pjackson28** [comment]
Strange as that is the same version of Chrome I'm using. Are you using my cleaned-up markup (unchanged) from https://github.com/wet-boew/wet-boew/issues/6416#issuecomment-68900166? Are you getting any errors in the console?

**Comment by josephbr** [comment]
Found these two errors in the console. Could they be related to the problem?
```
Resource interpreted as Image but transferred with MIME type application/javascript: "http://192.168.104.96/newsite_static/js/i18n/en.min.js".
Resource interpreted as Image but transferred with MIME type application/javascript: "http://192.168.104.96/newsite_static/js/deps/jquery.magnific-popup.min.js"
```

**Comment by pjackson28** [comment]
I don't think so as I get that elsewhere as well (for instance on the working examples page) and it doesn't seem to affect the functionality or display. Might be Chrome getting a little confused (harmlessly) when a JS file is dynamically loaded.

**Comment by pjackson28** [comment]
Can you post a screenshot of the issue that you are getting? Are you using the cleaned-up markup for your test?

**Comment by olegk1** [comment]
==Are you using my cleaned-up markup (unchanged) from #6416 (comment)?==
Yes, I am. Just checked it, and the issue is still there.

**Comment by pjackson28** [comment]
Please clarify exactly what is the issue and post a screenshot as well. I've downloaded the latest WET development release, used the cleaned-up code from above and tested on all available browsers and everything is behaving exactly as expected. Tabs are there when they are supposed to be and disappear when they are supposed to. Keyboard handling and transitions are behaving as expected and everything is displaying as it should be.

**Comment by josephbr** [comment]
Sorry, it seems I had the wrong page up. I checked the console again and this is the error I get on that page is: 

Chrome:
```
Uncaught TypeError: Cannot read property 'hasClass' of undefined 
wet-boew.min.js:11 
```

Firefox:
```
TypeError: c is undefined
wet-boew.min.js:11 
```

**Comment by olegk1** [comment]
Sorry, Paul, I confused you. Your cleaned up version does work now so I hope to use it on my pages as well expecting it to work. However. it doesn't work properly on Mac's browsers Firefox and Chrome. When you tab between tabs and ope/close buttons it skips tabs altogether. Could you please test it on Mac as well to reproduce it?

**Comment by pjackson28** [comment]
@josephbr Could you please try with a file from the unmin directory? It will make for much more meaningful error messages (i.e., with better line numbers).

@olegk1 I don't have access to Mac myself (except on SauceLabs) so I won't be able to do this custom test very easily. Could you please provide a step by step about what is happening on Mac and provide what is happening on PC so I can understand the difference?

**Comment by olegk1** [comment]
On Win tabbing works as it should with both Tab and Shft-Tab when I move from nested bars to open all-close all buttons and to tabs, and between the tabs, and back down to open all-close all buttons, and to nested bars. 

On Mac's Firefox and Safari; here's the order : 1) e.g starting point: focus is with the first nested collapsible bar--> 2) clicking Shift-Tab--> 3) focus is now with "close all" button--> 4) clicking Shift-Tab--> 5) focus is now with "open all" button 5) clicking Shift-Tab--> 5) focus is now with main collapsible bar. The issue: tabs are skipped when you try to Shift-Tab or just Tab between main collapsible bar and nested collapsible bars. When I tab down from main collapsible bar I expect focus to be on 2013-14 tab. Instead the focus is on "open all" button. Shift-Tab back doesn't bring focus to 2012-13 tab in order to navigate to its nested collapsible bars. 
But Crtl-Up brings focus from collapsible bars right to 2013-14 tab bypassing open all and close all and then using arrow right I can bring focus from 2013-14 tab to 2012-13 tab.

**Comment by pjackson28** [comment]
On Mac, what happens when you try to tab to other elements on the page? For instance, does it go to all links on the page? I have seen in the past where some browsers (particularly on Mac) that the default configuration is to tab only between form elements and a small number of other select elements (so links could not be tabbed to). Changing that option in the setting then would allow it behave like you expect.

Here is a snippet from http://www.webassign.net/manual/student_guide/t_a_osx_tab_config.htm:
"By default, Mac® OS X is not configured to use the TAB key to navigate to all items in your browser. If you are using Apple® Safari®, you also need to configure it to use TAB to navigate to all items on the page."

**Comment by olegk1** [comment]
Perfect! I've changed the Mac's browsers settings and now everything is fine - tabbing works between links as well. Thanks for the help!

**Comment by josephbr** [comment]
@pjackson28:

I used the non-minified version but it produced a different error for line 10409 (just updated WET again today with the latest development released):
```
TypeError: $elms is undefined
```

I went to that line number and saw this:
```
			if ( isInit && !isSmallView && $elms.hasClass( tabsAccordionClass ) ) {
				setTimeout(function() {
					$elms
						.removeAttr( "role" )
						.find( nestedTglPanelSelector ).removeAttr( "role" );
				}, 1 );
			}
```

**Comment by pjackson28** [comment]
Thanks! Found the issue. Will be fixed by PR #6434.
